plugins/python-build/scripts/add_miniconda.py

```diff
@@ -0,0 +1,261 @@
+  ;;
+""".strip()
+
+here = Path(__file__).resolve()
+out_dir: Path = here.parent.parent / "share" / "python-build"
+
+
+class StrEnum(str, Enum):
+    """Enum subclass which members are also instances of str
+    and directly comparable to strings. str type is forced at declaration.
+
+    Adapted from https://github.com/kissgyorgy/enum34-custom/blob/dbc89596761c970398701d26c6a5bbcfcf70f548/enum_custom.py#L100
+    (MIT license)
+    """
+    def __new__(cls, *args):
+        for arg in args:
+            if not isinstance(arg, str):
+                raise TypeError('Not text %s:' % arg)
+
+        return super(StrEnum, cls).__new__(cls, *args)
+
+    def __str__(self):
+        return str(self.value)
+
+
+class SupportedOS(StrEnum):
+    LINUX = "Linux"
+    MACOSX = "MacOSX"
+
+
+class SupportedArch(StrEnum):
+    PPC64LE = "ppc64le"
+    X86_64 = "x86_64"
+    X86 = "x86"
+
+
+class Suffix(StrEnum):
+    TWO = "2"
+    THREE = "3"
+
+
+class PyVersion(StrEnum):
+    PY27 = "py27"
+    PY36 = "py36"
+    PY37 = "py37"
+    PY38 = "py38"
+    PY39 = "py39"
+
+    def version(self):
+        first, *others = self.value[2:]
+        return f"{first}.{''.join(others)}"
+
+    def version_info(self):
+        return tuple(int(n) for n in self.version().split('.'))
+
+
+@total_ordering
+class VersionStr(str):
+    def info(self):
+        return tuple(int(n) for n in self.split('.'))
+
+    def __eq__(self, other):
+        return str(self) == str(other)
+
+    def __lt__(self, other):
+        if isinstance(other, VersionStr):
+            return self.info() < other.info()
+        raise ValueError("VersionStr can only be compared to other VersionStr")
+
+    @classmethod
+    def from_info(cls, version_info):
+        return VersionStr('.'.join(str(n) for n in version_info))
+
+    def __hash__(self):
+        return hash(str(self))
+
+
+class MinicondaVersion(NamedTuple):
+    suffix: Suffix
+    version_str: VersionStr
+
+    @classmethod
+    def from_str(cls, s):
+        miniconda_n, ver = s.split('-')
+        return MinicondaVersion(Suffix(miniconda_n[-1]), VersionStr(ver))
+
+    def to_filename(self):
+        return f"miniconda{self.suffix}-{self.version_str}"
+
+    def default_py_version(self):
+        if self.suffix == Suffix.TWO:
+            return PyVersion.PY27
+        else:
+            # https://docs.conda.io/projects/conda/en/latest/user-guide/tasks/manage-python.html
+            return PyVersion.PY36
+
+    def with_version_triple(self):
+        return MinicondaVersion(self.suffix, VersionStr.from_info(self.version_str.info()[:3]))
+
+
+class MinicondaSpec(NamedTuple):
+    version: MinicondaVersion
+    os: SupportedOS
+    arch: SupportedArch
+    md5: str
+    py_version: Optional[PyVersion] = None
+
+    @classmethod
+    def from_filestem(cls, stem, md5, py_version=None):
```

Comment:
Please consider using sha256sum in lieu of md5sum. The current code actually can validate MD5 checksum, however, it's just left for backward compatibility and no new code should use the legacy hashing algo.

Reply:
I used md5 because that's the checksum they have on the installer archive. If there's an equally accessible sha256 checksum I'll look into it.

Reply:
I still prefer to use SHA256 in general as MD5 is not considered as safe since awhile ago. However, I can understand that it's not possible to generate SHA256 without downloading remote contents in case the official doesn't provide it. It might be nice if generated file will have some shell script comment like `# FIXME: please replace this md5sum by sha256sum` maybe...? 🤔